samples/ClassroomLabs/Modules/Library/Tools/Readme.md

```diff
@@ -1,20 +1,138 @@
-1. Execute the Bulk Lab Creation script with the following commandline:
+1. Update the example CSV files to configure the resources to be created.  For additional labs, create additional lines in the CSV file.  The CSV files can be modified directly with Microsoft Excel.
+1. Import the functions in the script by typing:
+```powershell
+Import-Module LabCreationLibrary.psm1
+```
+
+## Examples
+The functions are generic and can be composed together to achieve different aims. In the following examples, we load the configuration information from a CSV file. The examples work the same if the information is loaded from a database. You need to substitute the first function with a database retrieving one.
+
+The full code for the example is immediately after the title in parenthesis.
+
+### Publish all the labs in a CSV file ([Examples/PublishAll.ps1](./Examples/PublishAll.ps1))
+
 ```powershell
-.\LabCreator.ps1 -CsvConfigFile hogwarts.csv
+".\hogwarts.csv" | Import-LabsCsv | Publish-Labs
 ```
+* `Import-LabsCsv` loads the configuration information from the csv file. It also loads schedule information for each lab from a separate file.
+* `Publish-Labs` publishes the labs and it is the natural end to all our pipelines. You can specify how many concurrent threads to use with the parameter `ThrottleLimit`.
+
+### Publish all the labs in a JSON file ([Examples/PublishAllJson.ps1](./Examples/PublishAllJson.ps1))
+You can store your configuration information in JSON form as well (i.e., file, Azure CosmoDb, other).
+
+```powershell
+Get-Content -Raw -Path "hogwarts.json" | ConvertFrom-Json | Publish-Labs
+```
+
+### Publish all the labs with a particular tag ([Examples/PublishByTag.ps1](./Examples/PublishByTag.ps1))
+
+```powershell
+".\hogwarts.csv"  | Import-LabsCsv | Select-Lab -SomeTags Maths, Science | Publish-Labs
+```
+
+In the example csv configuration file, the `Tags` column contains the emails of the professors that are giving a certain course. You can create the labs for a particular professor with:
+
+```powershell
+".\hogwarts.csv"  | Import-LabsCsv | Select-Lab -SomeTags bool1@hotmail.com | Publish-Labs
+```
+
+* `Select-Lab` is just a convenience function. You can achieve the same result by using any of the `Powershell` provided selection functions (i.e., `Where-Object`).
+ 
+### Publish one lab with a given id ([Examples/PublishById.ps1](./Examples/PublishById.ps1))
+
+```powershell
+".\hogwarts.csv"  | Import-LabsCsv | Select-Lab -Id id001 | Publish-Labs
+```
+
+### Publish one lab changing some configuration properties (i.e. a staging ResourceGroup)([Examples/PublishStaging.ps1](./Examples/PublishStaging.ps1))
+
+```powershell
+".\hogwarts.csv" `
+  | Import-LabsCsv `
+  | Select-Lab -Id id001 `
+  | Set-LabProperty -ResourceGroup Staging -MaxUsers 50 `
```

Comment:
I see how this is used now - is changing the resource group and/or the Lab Account the only times we would use this particular feature of overwriting the property across all the labs (with the same value)?  Does this get easier if we enable creating lab accounts as a separate function in the library?

Reply:
You can change any property for whatever set of labs you end up with in the pipeline. That is quite powerful as it supports many customer scenarios, not just the Staging one.